File: PirateLexer/Tokens/KeyWordService.cs
using PirateLexer.Enums;
using PirateLexer.Tokens.Interfaces;

namespace PirateLexer.Tokens;

/// <summary>
/// A class which checks for keywords.
/// </summary>
public class KeyWordService : IKeyWordService
{
    private string[] typeKeywords = new string[] { "var", "int", "float", "string", "char", "void" };

    private string[] controlKeywords = new string[] { "if", "else", "for", "to", "foreach", "in", "while", "func", "class", "new", "return" };


    public TokenType GetTypeKeyword(string idString)
    {
        if (typeKeywords.Contains(idString))
        {
            switch (idString)
            {
                case "var":
                    return TokenType.VAR;
                case "int":
                    return TokenType.INT;
                case "float":
                    return TokenType.FLOAT;
                case "string":
                    return TokenType.STRING;
                case "char":
                    return TokenType.CHAR;
                case "void":
                    return TokenType.VOID;
            }
            throw new NotImplementedException($"Type keyword, {idString} has not been implemented");
        }
        return TokenType.Empty;
    }

    public TokenType GetTokenControlKeyword(string idString)
    {
        if (controlKeywords.Contains(idString))
        {
            switch (idString)
            {
                case "if":
                    return TokenType.IF;

                case "else":
                    return TokenType.ELSE;

                case "for":
                    return TokenType.FOR;

                case "to":
                    return TokenType.TO;

                case "foreach":
                    return TokenType.FOREACH;

                case "in":
                    return TokenType.IN;

                case "while":
                    return TokenType.WHILE;

                case "func":
                    return TokenType.FUNC;

                case "class":
                    return TokenType.CLASS;

                case "new":
                    return TokenType.NEW;

                case "return":
                    return TokenType.RETURN;
            }
            throw new NotImplementedException($"Control keyword, {idString} has not been implemented");
        }
        return TokenType.Empty;
    }
}

File: Pirate.Lexer.F/KeyWordService.fs
namespace Pirate.Lexer

open Pirate.Lexer.Enums

type KeyWordService() =
    member _.typeKeyWords = ["var"; "int"; "float"; "string"; "char"; "void"]
    member _.controlKeyWords = [ "if"; "else"; "for"; "to"; "foreach"; "in"; "while"; "func"; "class"; "new"; "return" ]

    member this.GetTypeKeyword (idString:string) : TokenType =
        if this.typeKeyWords |> List.contains idString then
            match idString with
            | "var" -> VAR;
            | "int" -> INT;
            | "float" -> FLOAT;
            | "string" -> STRING;
            | "char" -> CHAR;
            | "void" -> VOID;
            | _ -> raise (System.NotImplementedException($"Type keyword, {idString} has not been implemented"));
        else
            TokenType.Empty;

    member this.GetTokenControlKeyword (idString:string) : TokenType =
        if this.controlKeyWords |> List.contains idString then
            match idString with
            | "if" -> IF;
            | "else" -> ELSE;
            | "for" -> FOR;
            | "to" -> TO;
            | "foreach" -> FOREACH;
            | "in" -> IN;
            | "while" -> WHILE;
            | "func" -> FUNC;
            | "class" -> CLASS;
            | "new" -> NEW;
            | "return" -> RETURN;
            | _ -> raise (System.NotImplementedException($"Control keyword, {idString} has not been implemented"));
        else
            TokenType.Empty;